File: lib/product/fixer.ml
type proposal = {
  id : string;
  description : string;
  edits : Yaml_cst.edit list;
  safe : bool;
}

let make description edits safe =
  let material =
    description
    ^ String.concat ""
        (List.map
           (fun (edit : Yaml_cst.edit) ->
             Printf.sprintf "%d:%d:%s" edit.start_byte edit.stop_byte
               edit.replacement)
           edits)
  in
  {
    id = "fix_" ^ String.sub (Sha256.digest_string material) 0 20;
    description;
    edits;
    safe;
  }

let replace_scalar ~cst:_ ~(scalar : Yaml_cst.scalar) ~replacement ~description
    =
  make description
    [
      {
        Yaml_cst.start_byte = scalar.Yaml_cst.span.start.byte;
        stop_byte = scalar.span.stop.byte;
        replacement;
      };
    ]
    true

let rec find_scalar reference = function
  | Yaml_cst.Scalar scalar when scalar.value = reference -> Some scalar
  | Scalar _ | Alias _ | Invalid _ -> None
  | Sequence (items, _) ->
      List.find_map
        (fun (item : Yaml_cst.sequence_item) ->
          find_scalar reference item.value)
        items
  | Flow_sequence (nodes, _) -> List.find_map (find_scalar reference) nodes
  | Mapping (entries, _) | Flow_mapping (entries, _) ->
      List.find_map
        (fun (entry : Yaml_cst.mapping_entry) ->
          match find_scalar reference entry.key_node with
          | Some _ as found -> found
          | None -> find_scalar reference entry.value)
        entries
  | Decorated decorated -> find_scalar reference decorated.value

let pin_dependency ~cst ~reference ~revision =
  let prefix =
    match String.rindex_opt reference '@' with
    | Some index -> String.sub reference 0 (index + 1)
    | None -> reference ^ "@"
  in
  match Option.bind (Yaml_cst.root cst) (find_scalar reference) with
  | None -> None
  | Some scalar ->
      Some
        (replace_scalar ~cst ~scalar ~replacement:(prefix ^ revision)
           ~description:("pin " ^ reference ^ " to " ^ revision))

let reduce_write_all ~cst ~unused_capabilities =
  if
    not
      (List.for_all
         (fun capability -> List.mem capability unused_capabilities)
         [ Ir.Repository_write; Ir.Token_write ])
  then None
  else
    match Option.bind (Yaml_cst.root cst) (find_scalar "write-all") with
    | None -> None
    | Some scalar ->
        Some
          (replace_scalar ~cst ~scalar ~replacement:"read-all"
             ~description:
               "reduce write-all after proving repository and token write \
                grants unused")

let substring_index ~needle value =
  let needle_length = String.length needle
  and value_length = String.length value in
  let rec search index =
    if index + needle_length > value_length then None
    else if String.sub value index needle_length = needle then Some index
    else search (index + 1)
  in
  if needle_length = 0 then None else search 0

let exactly_once ~needle value =
  match substring_index ~needle value with
  | None -> None
  | Some first ->
      let after = first + String.length needle in
      let suffix = String.sub value after (String.length value - after) in
      if Option.is_some (substring_index ~needle suffix) then None
      else Some first

let valid_environment_name value =
  let valid_start = function
    | 'A' .. 'Z' | 'a' .. 'z' | '_' -> true
    | _ -> false
  and valid_rest = function
    | 'A' .. 'Z' | 'a' .. 'z' | '0' .. '9' | '_' -> true
    | _ -> false
  in
  String.length value > 0
  && valid_start value.[0]
  && String.sub value 1 (String.length value - 1) |> String.for_all valid_rest

let rec find_run expression = function
  | Yaml_cst.Mapping (entries, _) | Flow_mapping (entries, _) -> (
      let direct =
        if Option.is_some (Yaml_cst.mapping_find "env" entries) then None
        else
          entries
          |> List.find_map (fun (entry : Yaml_cst.mapping_entry) ->
              if entry.key.value <> "run" then None
              else
                match entry.value with
                | Scalar scalar
                  when scalar.style = Yaml_cst.Plain
                       && scalar.raw = scalar.value
                       && Option.is_some
                            (substring_index ~needle:expression scalar.value) ->
                    Some (entry, scalar)
                | _ -> None)
      in
      match direct with
      | Some _ as value -> value
      | None ->
          entries
          |> List.find_map (fun (entry : Yaml_cst.mapping_entry) ->
              match find_run expression entry.value with
              | Some _ as value -> value
              | None -> find_run expression entry.key_node))
  | Sequence (items, _) ->
      items
      |> List.find_map (fun (item : Yaml_cst.sequence_item) ->
          find_run expression item.value)
  | Flow_sequence (items, _) -> List.find_map (find_run expression) items
  | Decorated decorated -> find_run expression decorated.value
  | Scalar _ | Alias _ | Invalid _ -> None

let line_insertion cst stop_byte =
  let source = cst.Yaml_cst.source in
  let rec find index =
    if index >= String.length source then (String.length source, false)
    else
      match source.[index] with
      | '\n' -> (index + 1, true)
      | '\r' ->
          if index + 1 < String.length source && source.[index + 1] = '\n' then
            (index + 2, true)
          else (index + 1, true)
      | _ -> find (index + 1)
  in
  find stop_byte

let bind_expression_to_environment ~cst ~shell ~expression ~name =
  if
    (not (valid_environment_name name))
    || not
         (Util.starts_with ~prefix:"${{" expression
         || Util.starts_with ~prefix:"$[[" expression)
  then None
  else
    match Option.bind (Yaml_cst.root cst) (find_run expression) with
    | None -> None
    | Some (entry, scalar) -> (
        match exactly_once ~needle:expression scalar.value with
        | None -> None
        | Some relative ->
            let variable =
              match shell with
              | Script_adapter.Posix | Bash -> Some ("\"${" ^ name ^ "}\"")
              | PowerShell -> Some ("\"$env:" ^ name ^ "\"")
              | Cmd -> Some ("\"%" ^ name ^ "%\"")
              | Python | Unknown_shell _ -> None
            in
            Option.map
              (fun variable ->
                let insertion, had_newline =
                  line_insertion cst scalar.span.stop.byte
                and newline =
                  match cst.newline with
                  | `CrLf -> "\r\n"
                  | `Cr -> "\r"
                  | `Lf | `None -> "\n"
                and indent =
                  String.make (max 0 (entry.key.span.start.column - 1)) ' '
                in
                let env_block =
                  (if had_newline then "" else newline)
                  ^ indent ^ "env:" ^ newline ^ indent ^ "  " ^ name ^ ": "
                  ^ expression ^ newline
                in
                make
                  ("bind " ^ expression ^ " through environment variable "
                 ^ name)
                  [
                    {
                      Yaml_cst.start_byte = scalar.span.start.byte + relative;
                      stop_byte =
                        scalar.span.start.byte + relative
                        + String.length expression;
                      replacement = variable;
                    };
                    {
                      Yaml_cst.start_byte = insertion;
                      stop_byte = insertion;
                      replacement = env_block;
                    };
                  ]
                  true)
              variable)

let apply ~cst proposal =
  if not proposal.safe then
    Error "unsafe proposals cannot be applied automatically"
  else Yaml_cst.apply_edits cst proposal.edits

let combine proposals =
  if proposals = [] then Error "at least one fix proposal is required"
  else if List.exists (fun proposal -> not proposal.safe) proposals then
    Error "unsafe proposals cannot be combined for automatic application"
  else
    let edits =
      proposals
      |> List.concat_map (fun proposal -> proposal.edits)
      |> List.sort (fun left right ->
          match Int.compare left.Yaml_cst.start_byte right.start_byte with
          | 0 -> Int.compare left.stop_byte right.stop_byte
          | comparison -> comparison)
    in
    let rec overlap = function
      | left :: (right :: _ as rest) ->
          if left.Yaml_cst.stop_byte > right.start_byte then true
          else overlap rest
      | _ -> false
    in
    if overlap edits then Error "fix proposals contain overlapping edits"
    else
      Ok
        (make
           (proposals
           |> List.map (fun proposal -> proposal.description)
           |> Util.deduplicate_strings |> String.concat "; ")
           edits true)

let diff_lines source =
  let lines = String.split_on_char '\n' source in
  match List.rev lines with
  | "" :: rest -> List.rev rest
  | _ -> lines

let unified_diff ~path ~before ~after =
  if before = after then ""
  else
    let before_lines = diff_lines before and after_lines = diff_lines after in
    let buffer =
      Buffer.create (String.length before + String.length after + 128)
    in
    Printf.bprintf buffer "--- %s\n+++ %s\n@@ -1,%d +1,%d @@\n"
      (Util.normalize_slashes path)
      (Util.normalize_slashes path)
      (List.length before_lines) (List.length after_lines);
    List.iter
      (fun line -> Buffer.add_string buffer ("-" ^ line ^ "\n"))
      before_lines;
    List.iter
      (fun line -> Buffer.add_string buffer ("+" ^ line ^ "\n"))
      after_lines;
    Buffer.contents buffer

let to_json proposal =
  Json.Object
    [
      ("description", Json.String proposal.description);
      ( "edits",
        Json.Array
          (List.map
             (fun (edit : Yaml_cst.edit) ->
               Json.Object
                 [
                   ("replacement", Json.String edit.replacement);
                   ("start_byte", Json.Int edit.start_byte);
                   ("stop_byte", Json.Int edit.stop_byte);
                 ])
             proposal.edits) );
      ("id", Json.String proposal.id);
      ("safe", Json.Bool proposal.safe);
    ]

File: src/product/fixer.rs
use crate::domain::Capability;
use crate::foundation::{Budget, JsonValue, normalize_slashes, sha256_hex};
use crate::syntax::{Edit, MappingEntry, ScalarStyle, YamlDocument, YamlNode};
use std::collections::{BTreeMap, BTreeSet};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum FixShell {
    Posix,
    Bash,
    PowerShell,
    Cmd,
    Python,
    Unknown,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct FixProposal {
    pub id: String,
    pub description: String,
    pub edits: Vec<Edit>,
    pub safe: bool,
}

impl FixProposal {
    fn new(description: impl Into<String>, mut edits: Vec<Edit>, safe: bool) -> Self {
        use std::fmt::Write as _;
        let description = description.into();
        edits.sort_by(|left, right| {
            left.start_byte
                .cmp(&right.start_byte)
                .then(left.stop_byte.cmp(&right.stop_byte))
                .then(left.replacement.cmp(&right.replacement))
        });
        let mut material = description.clone();
        for edit in &edits {
            let _ = write!(
                material,
                "{}:{}:{}",
                edit.start_byte, edit.stop_byte, edit.replacement
            );
        }
        let digest = sha256_hex(material);
        Self {
            id: format!("fix_{}", digest.get(..20).unwrap_or(&digest)),
            description,
            edits,
            safe,
        }
    }

    #[must_use]
    pub fn replace_span(
        start_byte: usize,
        stop_byte: usize,
        replacement: impl Into<String>,
        description: impl Into<String>,
    ) -> Self {
        Self::new(
            description,
            vec![Edit::replace(start_byte, stop_byte, replacement)],
            true,
        )
    }

    #[must_use]
    pub fn pin_dependency(
        document: &YamlDocument,
        reference: &str,
        revision: &str,
    ) -> Option<Self> {
        let node = find_scalar(document.root()?, reference)?;
        let prefix = reference.rfind('@').map_or_else(
            || format!("{reference}@"),
            |index| reference[..=index].to_owned(),
        );
        Some(Self::replace_span(
            node.span().start.byte,
            node.span().stop.byte,
            format!("{prefix}{revision}"),
            format!("pin {reference} to {revision}"),
        ))
    }

    #[must_use]
    pub fn reduce_write_all(
        document: &YamlDocument,
        unused_capabilities: &[Capability],
    ) -> Option<Self> {
        if ![Capability::RepositoryWrite, Capability::TokenWrite]
            .iter()
            .all(|capability| unused_capabilities.contains(capability))
        {
            return None;
        }
        let node = find_scalar(document.root()?, "write-all")?;
        Some(Self::replace_span(
            node.span().start.byte,
            node.span().stop.byte,
            "read-all",
            "reduce write-all after proving repository and token write grants unused",
        ))
    }

    #[must_use]
    pub fn bind_expression_to_environment(
        document: &YamlDocument,
        shell: FixShell,
        expression: &str,
        name: &str,
    ) -> Option<Self> {
        if !environment_name(name)
            || !(expression.starts_with("${{") || expression.starts_with("$[["))
        {
            return None;
        }
        let (entry, scalar) = find_run(document.root()?, expression)?;
        if scalar.scalar_style() != Some(ScalarStyle::Plain) {
            return None;
        }
        let command = scalar.scalar()?;
        let relative = exactly_once(command, expression)?;
        let variable = match shell {
            FixShell::Posix | FixShell::Bash => format!("\"${{{name}}}\""),
            FixShell::PowerShell => format!("\"$env:{name}\""),
            FixShell::Cmd => format!("\"%{name}%\""),
            FixShell::Python | FixShell::Unknown => return None,
        };
        let (insertion, has_newline) = line_insertion(document.print(), scalar.span().stop.byte);
        let newline = if document.print().contains("\r\n") {
            "\r\n"
        } else if document.print().contains('\r') {
            "\r"
        } else {
            "\n"
        };
        let indent = " ".repeat(entry.key_span.start.column.saturating_sub(1) as usize);
        let block = format!(
            "{}{indent}env:{newline}{indent}  {name}: {expression}{newline}",
            if has_newline { "" } else { newline }
        );
        Some(Self::new(
            format!("bind {expression} through environment variable {name}"),
            vec![
                Edit::replace(
                    scalar.span().start.byte + relative,
                    scalar.span().start.byte + relative + expression.len(),
                    variable,
                ),
                Edit::replace(insertion, insertion, block),
            ],
            true,
        ))
    }

    /// Combine safe, non-overlapping proposals into one atomic transaction.
    ///
    /// # Errors
    /// Rejects empty, unsafe, or overlapping proposal sets.
    pub fn combine(proposals: &[Self]) -> Result<Self, String> {
        if proposals.is_empty() {
            return Err("at least one fix proposal is required".to_owned());
        }
        if proposals.iter().any(|proposal| !proposal.safe) {
            return Err("unsafe proposals cannot be combined for automatic application".to_owned());
        }
        let mut edits: Vec<_> = proposals
            .iter()
            .flat_map(|proposal| proposal.edits.iter().cloned())
            .collect();
        edits.sort_by(|left, right| {
            left.start_byte
                .cmp(&right.start_byte)
                .then(left.stop_byte.cmp(&right.stop_byte))
        });
        if edits.windows(2).any(|pair| {
            let left = &pair[0];
            let right = &pair[1];
            left.stop_byte > right.start_byte
                || (left.start_byte == left.stop_byte
                    && right.start_byte == right.stop_byte
                    && left.start_byte == right.start_byte)
        }) {
            return Err("fix proposals contain overlapping edits".to_owned());
        }
        let descriptions: BTreeSet<_> = proposals
            .iter()
            .map(|proposal| proposal.description.clone())
            .collect();
        Ok(Self::new(
            descriptions.into_iter().collect::<Vec<_>>().join("; "),
            edits,
            true,
        ))
    }

    /// Apply a safe transaction and require the result to parse cleanly.
    ///
    /// # Errors
    /// Rejects unsafe/out-of-range edits or a malformed resulting YAML document.
    pub fn apply(&self, document: &YamlDocument) -> Result<String, String> {
        if !self.safe {
            return Err("unsafe proposals cannot be applied automatically".to_owned());
        }
        let output = document.apply_edits(&self.edits)?;
        let reparsed = YamlDocument::parse(document.file(), &output, Budget::default());
        if reparsed.root().is_none() || !reparsed.invalid_regions().is_empty() {
            return Err("fixed document did not reparse cleanly".to_owned());
        }
        Ok(output)
    }

    /// Apply and ask a caller-owned analyzer to re-prove the semantic safety condition.
    ///
    /// # Errors
    /// Returns the edit or proof failure without mutating a file.
    pub fn apply_verified(
        &self,
        document: &YamlDocument,
        prove: impl FnOnce(&str) -> Result<(), String>,
    ) -> Result<String, String> {
        let output = self.apply(document)?;
        prove(&output)?;
        Ok(output)
    }

    #[must_use]
    pub fn unified_diff(&self, path: &str, before: &str, after: &str) -> String {
        if before == after {
            return String::new();
        }
        let before_lines = diff_lines(before);
        let after_lines = diff_lines(after);
        let path = normalize_slashes(path);
        let mut output = format!(
            "--- {path}\n+++ {path}\n@@ -1,{} +1,{} @@\n",
            before_lines.len(),
            after_lines.len()
        );
        for line in before_lines {
            output.push('-');
            output.push_str(line);
            output.push('\n');
        }
        for line in after_lines {
            output.push('+');
            output.push_str(line);
            output.push('\n');
        }
        output
    }

    #[must_use]
    pub fn to_json(&self) -> JsonValue {
        JsonValue::Object(BTreeMap::from([
            (
                "description".to_owned(),
                JsonValue::String(self.description.clone()),
            ),
            (
                "edits".to_owned(),
                JsonValue::Array(
                    self.edits
                        .iter()
                        .map(|edit| {
                            JsonValue::Object(BTreeMap::from([
                                (
                                    "replacement".to_owned(),
                                    JsonValue::String(edit.replacement.clone()),
                                ),
                                (
                                    "start_byte".to_owned(),
                                    JsonValue::Integer(
                                        i64::try_from(edit.start_byte).unwrap_or(i64::MAX),
                                    ),
                                ),
                                (
                                    "stop_byte".to_owned(),
                                    JsonValue::Integer(
                                        i64::try_from(edit.stop_byte).unwrap_or(i64::MAX),
                                    ),
                                ),
                            ]))
                        })
                        .collect(),
                ),
            ),
            ("id".to_owned(), JsonValue::String(self.id.clone())),
            ("safe".to_owned(), JsonValue::Boolean(self.safe)),
        ]))
    }
}

fn find_scalar<'a>(node: &'a YamlNode, wanted: &str) -> Option<&'a YamlNode> {
    if node.scalar() == Some(wanted) {
        return Some(node);
    }
    node.mapping()
        .into_iter()
        .flatten()
        .find_map(|entry| find_scalar(&entry.value, wanted))
        .or_else(|| {
            node.sequence()
                .into_iter()
                .flatten()
                .find_map(|item| find_scalar(item, wanted))
        })
}

fn find_run<'a>(node: &'a YamlNode, expression: &str) -> Option<(&'a MappingEntry, &'a YamlNode)> {
    if let Some(entries) = node.mapping() {
        if entries.iter().all(|entry| entry.key != "env")
            && let Some(found) = entries.iter().find_map(|entry| {
                (entry.key == "run"
                    && entry
                        .value
                        .scalar()
                        .is_some_and(|value| value.contains(expression)))
                .then_some((entry, &entry.value))
            })
        {
            return Some(found);
        }
        if let Some(found) = entries
            .iter()
            .find_map(|entry| find_run(&entry.value, expression))
        {
            return Some(found);
        }
    }
    node.sequence()
        .into_iter()
        .flatten()
        .find_map(|item| find_run(item, expression))
}

fn exactly_once(haystack: &str, needle: &str) -> Option<usize> {
    let first = haystack.find(needle)?;
    (!haystack[first + needle.len()..].contains(needle)).then_some(first)
}

fn environment_name(value: &str) -> bool {
    let mut bytes = value.bytes();
    bytes
        .next()
        .is_some_and(|byte| byte.is_ascii_alphabetic() || byte == b'_')
        && bytes.all(|byte| byte.is_ascii_alphanumeric() || byte == b'_')
}

fn line_insertion(source: &str, stop: usize) -> (usize, bool) {
    let suffix = &source[stop.min(source.len())..];
    for (relative, character) in suffix.char_indices() {
        match character {
            '\n' => return (stop + relative + 1, true),
            '\r' => {
                let width = usize::from(suffix.as_bytes().get(relative + 1) == Some(&b'\n')) + 1;
                return (stop + relative + width, true);
            }
            _ => {}
        }
    }
    (source.len(), false)
}

fn diff_lines(source: &str) -> Vec<&str> {
    let mut lines: Vec<_> = source.split('\n').collect();
    if lines.last() == Some(&"") {
        let _ = lines.pop();
    }
    lines
}

#[cfg(test)]
mod tests {
    use super::*;

    fn document(source: &str) -> YamlDocument {
        YamlDocument::parse("fix.yml", source, Budget::default())
    }

    #[test]
    fn lexical_fix_helpers_observe_every_boundary() {
        assert_eq!(
            exactly_once("prefix TOKEN suffix", "TOKEN"),
            Some("prefix ".len())
        );
        assert_eq!(exactly_once("TOKEN TOKEN", "TOKEN"), None);
        assert_eq!(exactly_once("missing", "TOKEN"), None);

        for valid in ["A", "_A", "A0", "name_with_digits_42"] {
            assert!(environment_name(valid), "valid environment name {valid:?}");
        }
        for invalid in ["", "0A", "-A", "A-B", "A B", "Ä"] {
            assert!(
                !environment_name(invalid),
                "invalid environment name {invalid:?}"
            );
        }

        assert_eq!(
            line_insertion("run: x\nnext: y\n", "run: x".len()),
            ("run: x\n".len(), true)
        );
        assert_eq!(
            line_insertion("run: value suffix\nnext: y\n", "run: ".len()),
            ("run: value suffix\n".len(), true)
        );
        assert_eq!(
            line_insertion("run: x\r\nnext: y\r\n", "run: x".len()),
            ("run: x\r\n".len(), true)
        );
        assert_eq!(
            line_insertion("run: value suffix\r\nnext: y\r\n", "run: ".len()),
            ("run: value suffix\r\n".len(), true)
        );
        assert_eq!(
            line_insertion("run: x\rnext: y\r", "run: x".len()),
            ("run: x\r".len(), true)
        );
        assert_eq!(
            line_insertion("run: x", "run: x".len()),
            ("run: x".len(), false)
        );
        assert_eq!(
            line_insertion("run: x", usize::MAX),
            ("run: x".len(), false)
        );

        assert_eq!(diff_lines(""), Vec::<&str>::new());
        assert_eq!(diff_lines("one"), ["one"]);
        assert_eq!(diff_lines("one\n"), ["one"]);
        assert_eq!(diff_lines("one\ntwo"), ["one", "two"]);
        assert_eq!(diff_lines("one\n\n"), ["one", ""]);
    }

    #[test]
    fn run_lookup_requires_a_plain_unique_expression_outside_env() {
        let nested =
            document("jobs:\n  build:\n    steps:\n      - run: echo ${{ inputs.value }}\n");
        let (entry, scalar) = find_run(nested.root().expect("nested root"), "${{ inputs.value }}")
            .expect("nested run");
        assert_eq!(entry.key, "run");
        assert_eq!(scalar.scalar(), Some("echo ${{ inputs.value }}"));

        let environment = document("env:\n  TOKEN: ${{ inputs.value }}\nrun: echo safe\n");
        assert!(
            find_run(
                environment.root().expect("environment root"),
                "${{ inputs.value }}"
            )
            .is_none()
        );
        let absent = document("run: echo safe\n");
        assert!(find_run(absent.root().expect("absent root"), "${{ inputs.value }}").is_none());
    }

    #[test]
    fn binding_contract_covers_shells_rejections_and_line_endings() {
        let cases = [
            (FixShell::Posix, "\"${INPUT_VALUE}\""),
            (FixShell::Bash, "\"${INPUT_VALUE}\""),
            (FixShell::PowerShell, "\"$env:INPUT_VALUE\""),
            (FixShell::Cmd, "\"%INPUT_VALUE%\""),
        ];
        for (shell, replacement) in cases {
            let source = "steps:\n  - run: echo ${{ inputs.value }}\n";
            let proposal = FixProposal::bind_expression_to_environment(
                &document(source),
                shell,
                "${{ inputs.value }}",
                "INPUT_VALUE",
            )
            .expect("supported binding");
            let output = proposal.apply(&document(source)).expect("binding applies");
            assert!(output.contains(&format!("run: echo {replacement}")));
            assert!(output.contains("    env:\n      INPUT_VALUE: ${{ inputs.value }}\n"));
        }

        let crlf = "steps:\r\n  - run: echo $[[ inputs.value ]]\r\n";
        let proposal = FixProposal::bind_expression_to_environment(
            &document(crlf),
            FixShell::Bash,
            "$[[ inputs.value ]]",
            "INPUT_VALUE",
        )
        .expect("GitLab expression binding");
        assert!(
            proposal
                .apply(&document(crlf))
                .expect("CRLF binding")
                .contains("    env:\r\n      INPUT_VALUE: $[[ inputs.value ]]\r\n")
        );

        let plain = document("run: echo ${{ inputs.value }}\n");
        for shell in [FixShell::Python, FixShell::Unknown] {
            assert!(
                FixProposal::bind_expression_to_environment(
                    &plain,
                    shell,
                    "${{ inputs.value }}",
                    "INPUT_VALUE",
                )
                .is_none()
            );
        }
        for (expression, name) in [
            ("inputs.value", "INPUT_VALUE"),
            ("${{ inputs.value }}", "0INVALID"),
        ] {
            assert!(
                FixProposal::bind_expression_to_environment(
                    &plain,
                    FixShell::Bash,
                    expression,
                    name,
                )
                .is_none()
            );
        }
        for source in [
            "run: echo ${{ inputs.value }} ${{ inputs.value }}\n",
            "run: 'echo ${{ inputs.value }}'\n",
            "env:\n  INPUT_VALUE: existing\nrun: echo ${{ inputs.value }}\n",
        ] {
            assert!(
                FixProposal::bind_expression_to_environment(
                    &document(source),
                    FixShell::Bash,
                    "${{ inputs.value }}",
                    "INPUT_VALUE",
                )
                .is_none()
            );
        }
    }

    #[test]
    fn transaction_contract_rejects_each_unsafe_shape_and_reproves_output() {
        let insertion = |at, replacement: &str| {
            FixProposal::new("insert", vec![Edit::replace(at, at, replacement)], true)
        };
        assert!(FixProposal::combine(&[]).is_err());
        let unsafe_proposal = FixProposal {
            id: "unsafe".to_owned(),
            description: "unsafe".to_owned(),
            edits: vec![Edit::replace(0, 0, "key: value\n")],
            safe: false,
        };
        assert!(FixProposal::combine(std::slice::from_ref(&unsafe_proposal)).is_err());
        assert!(unsafe_proposal.apply(&document("key: value\n")).is_err());

        let replacing = FixProposal::replace_span(0, "key".len(), "name", "replace key");
        assert!(FixProposal::combine(&[replacing.clone(), insertion("k".len(), "x")]).is_err());
        assert!(FixProposal::combine(&[insertion(0, "a"), insertion(0, "b")]).is_err());
        assert!(FixProposal::combine(&[insertion(0, "a"), insertion(1, "b")]).is_ok());
        assert!(
            FixProposal::combine(&[
                FixProposal::replace_span(0, "a".len(), "A", "replace first"),
                FixProposal::replace_span("a".len(), "ab".len(), "B", "replace second",),
            ])
            .is_ok()
        );

        let source = "key: value\n";
        let valid = FixProposal::replace_span(0, "key".len(), "name", "replace key");
        let mut proof_input = String::new();
        let output = valid
            .apply_verified(&document(source), |candidate| {
                proof_input = candidate.to_owned();
                Ok(())
            })
            .expect("reproved fix");
        assert_eq!(output, "name: value\n");
        assert_eq!(proof_input, output);
        assert!(
            valid
                .apply_verified(&document(source), |_| Err("proof rejected".to_owned()))
                .is_err()
        );

        let empty = FixProposal::replace_span(0, source.len(), "", "empty document");
        assert!(empty.apply(&document(source)).is_err());
        let malformed =
            FixProposal::replace_span(0, source.len(), "key: [\n", "malformed document");
        assert!(malformed.apply(&document(source)).is_err());
    }
}
